### ISSUE TITLE
Interrupted tests passed SIGKILL immediately by process-wrapper

### ISSUE THREAD

**Issue by pauldraper** [bazelbuild/bazel#14733]
### Description of the problem / feature request:

When bazel test is interrupted (SIGINT, SIGTEM), it sends SIGKILL to the test process, if process-wrapper is used.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

**WORKSPACE.bazel**

```bzl
```

**BUILD.bazel**

```bzl
sh_test(
    name = "example",
    srcs = ["example.sh"],
    tags = ["no-sandbox"],
)
```

**example.sh**

```sh
#!/usr/bin/env bash
trap 'echo DONE' EXIT
sleep 1000
```

Run

```sh
bazel test --test_output=streamed :example
```

When the test is running, sent it SIGINT/SIGTERM.

Notice that "DONE" is not printed.

### What operating system are you running Bazel on?

Ubuntu 20.04

### What's the output of `bazel info release`?

```
release 5.0.0
```

### Any other information, logs, or outputs that you want to share?

It is important for tests to received SIGINT/SIGTERM on interruption so that they can cleanup. Docker containers, VMs, etc.

---

I have traced this back to the process-wrapper exectuable. When the Linux sandbox is used, this problem does not exist. (That's why no-sandbox tag was necessary.)

```sh
"$(bazel info install_base)"/process-wrapper --timeout=300 --kill_delay=15 ./example.sh
```

This wrapper passes SIGKILL to the child, when it receives SIGINT/SIGTERM.

---

I can't emphasize how surprised I am at this behavior, especially by a tool (process-wrapper) designed *specifically* around process management/signal handling, and with flags like seem they were created for this :/

**Comment by pauldraper** [comment]
Actually, it appears that process-wrapper does have this option: `-g/--graceful_sigterm`. But it isn't being provided.

**Comment by pauldraper** [comment]
It makes sense to add `-g` to process-wrapper (or make it always happen).

**Comment by milancpp** [comment]
I am hitting this problem as well, and it is forcing me to temporarily switch the target of a large integration test to sh_binary from sh_test whenever I debug something using this test.

Tests need to be able to clean up after them, and Bazel is making this impossible. Immediately sending SIGKILL is a really bad choice, any process wrapper-type solution I've seen before had some kind of graceful interval during which it allows SIGTERM handlers to run.

**Comment by fmeum** [comment]
@larsrc-google Do you see anything preventing us from just passing `-g` to the process-wrapper?

**Comment by fmeum** [comment]
I just found https://github.com/bazelbuild/bazel/commit/2bb3b73d5739fd9cae152a3626b4a6c203d8cef6. Have you tried adding `tags = ["supports-graceful-termination"]` to your test?

**Comment by milancpp** [comment]
> I just found [2bb3b73](https://github.com/bazelbuild/bazel/commit/2bb3b73d5739fd9cae152a3626b4a6c203d8cef6). Have you tried adding `tags = ["supports-graceful-termination"]` to your test?

Good find! I tried with my sh_binary/sh_test but it doesn't seem to have any effect here @fmeum

**Comment by fmeum** [comment]
@milancpp Do you have a reproducer you can share? If that tag is not working as expected, that's a bug we should fix.

**Comment by fmeum** [comment]
@milancpp I can reproduce this with `bazel run`, but not with `bazel test`. `bazel run`ning a test is essentially best effort and doesn't use sandboxing - something like `bazel test ... --test_output=streamed -t-` is preferred.

**Comment by milancpp** [comment]
> @milancpp I can reproduce this with `bazel run`, but not with `bazel test`. `bazel run`ning a test is essentially best effort and doesn't use sandboxing - something like `bazel test ... --test_output=streamed -t-` is preferred.

That is very unfortunate, at least I and a couple of colleagues regularly use `run` on tests - basically whenever I am doing test-driven development. `test` is for when you just want to see the results and failures - not any intermediate output.

In this case, the test happens to be spawning containers and configure virtual networks. Not cleaning up means leaving these expensive entities around, bogging down the system :/

**Comment by fmeum** [comment]
@milancpp Is there a reason for why `bazel test //the:test --test_output=streamed -t-` doesn't work for you? Since `bazel run` doesn't use the sandbox and also doesn't perfectly mimic the `test` environment in other aspects, you may run into differing behavior.

**Comment by milancpp** [comment]
> @milancpp Is there a reason for why `bazel test //the:test --test_output=streamed -t-` doesn't work for you? Since `bazel run` doesn't use the sandbox and also doesn't perfectly mimic the `test` environment in other aspects, you may run into differing behavior.

I would argue that there shouldn't be differing behavior between `test`ing and `run`ing a test target, because both involve executing the same code. As mentioned above, using one or the other is a question of how the test is being used - test-driven development or classic test execution (e.g. in CI).

In my case, where missing cleanup because a fellow developer happened to use the wrong verb is not an option, I will probably workaround it by making the test a sh_binary and adding an additional sh_test target that executes the binary in CI. Thanks for the support @fmeum!

**Comment by fmeum** [comment]
> I would argue that there shouldn't be differing behavior between testing and runing a test target, because both involve executing the same code.

`test` means run in a test runner (e.g. to enforce timeouts, collect XML reports, ...), so I think it's unlikely that `run` can ever provide a 100% identical environment. 

> In my case, where missing cleanup because a fellow developer happened to use the wrong verb is not an option, I will probably workaround it by making the test a sh_binary and adding an additional sh_test target that executes the binary in CI.

@lberki Since I have been using `bazel run` on tests for a while before learning about the difference between it and `bazel test`, I wonder how situations like this are prevented at Google. Could we make running tests "the right way" simpler and also draw more attention to the difference?

**Comment by lberki** [comment]
Yes, it was a long discussion until we settled on the current semantics of `bazel run` for tests :)

It's due to a combination of three things:
1. Ease of implementation (it wouldn't have been easy to teach `bazel run` to run tests in a sandbox, etc., which is necessary to faithfully emulate the environment the test is running in)
2. Lack of need (as you say above, `--test_output=streamed` was deemed to be an adequate replacement for `bazel run` in cases where one wants to stream the output of the test to the terminal)
3. The very prevalent pattern of test binaries having a mode (usually activated by a command line flag) where they update their own golden files (which requires write access to the source tree and knowing where it is). This is not an endorsement of golden tests, merely a statement that there are a lot of those in our monorepo.

In what way is `--test_output=streamed` inadequate?

**Comment by milancpp** [comment]
@fmeum @lberki is there a way to disable the `run` verb for certain tests? Maybe a tag `no-run-allowed` that would translate to a friendly error when trying to `bazel run` it?

I don't have an issue with using `test --test_output=streamed` but I would like to prevent unintended left-overs on other developers machines when they innocently try to run the test without knowing about this limitation.

**Comment by lberki** [comment]
Can you do it in the test itself? The presence of `BUILD_WORKSPACE_DIRECTORY` / `BUILD_WORKING_DIRECTORY` environment variables during the test run is a good signal that it's run under `bazel run`.

**Comment by milancpp** [comment]
>Can you do it in the test itself? The presence of BUILD_WORKSPACE_DIRECTORY / BUILD_WORKING_DIRECTORY environment variables during the test run is a good signal that it's run under bazel run

Finally found time to try this out to detect `run` and error out with a good description. Works well! @lberki thanks for the suggestion, now `supports-graceful-termination` is usable for my use case.

**Comment by lberki** [comment]
@milancpp I'm glad I could help :)
